File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode balanceBST(TreeNode root) {
        //traverse using inorder and make a linkedList, which is sorted
        List<Integer> sortedList = new LinkedList<Integer>();
        inorder(root, sortedList);
        
        return createBST(sortedList, 0, sortedList.size()-1);
    }
    private void inorder(TreeNode root, List<Integer> list){
        if(root== null){
            return;
        }
        inorder(root.left, list);
        list.add(root.val);
        inorder(root.right, list);
    }
    
    private TreeNode createBST(List<Integer> list, int start, int end){
        if(start>end){
            return null;//leaf node child
        }
        int mid = (start+end)/2;
        TreeNode root = new TreeNode(list.get(mid), createBST(list, start, mid-1), createBST(list, mid+1, end));
        return root;
    }
}

File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun balanceBST(root: TreeNode?): TreeNode? {
        val list = mutableListOf<Int>()
        inorder(root, list);
        return createBST(list, 0, list.size - 1);
    }

    private fun inorder(root: TreeNode?, list: MutableList<Int>) {
        if (root == null) {
            return;
        }
        inorder(root.left, list);
        list.add(
            root.`val`
        );
        inorder(root.right, list);
    }

    private fun createBST(list: MutableList<Int>, start: Int, end: Int): TreeNode? {
        if (start > end) {
            return null;//leaf node child
        }
        val mid = (start + end) / 2;
        val root = TreeNode(list[mid])
        root.left = createBST(list, start, mid - 1)
        root.right = createBST(list, mid + 1, end)
        return root;
    }
}